grammar.js

```diff
@@ -998,6 +1027,146 @@ module.exports = grammar({
       'end',
     ),
 
+    lib_def: $ => seq(
+      'lib',
+      field('name', $.constant, $.generic_type),
+      optional($._lib_statements),
+      'end',
+    ),
+
+    top_level_fun_def: $ => {
+      const params = seq(
+        '(', optional(field('params', alias($.fun_param_list, $.param_list))), ')',
+      )
+      const real_name = seq('=',
+        field('real_name', choice($.identifier, $.constant, $.string)),
+      )
+      const return_type = field('type', seq(/[ \t]:\s/, $._bare_type))
+
+      return seq(
+        prec.right(seq(
+          'fun',
+          field('name', $.identifier),
+          optional(real_name),
+          optional(params),
+          optional(return_type),
+        )),
+        field('body', optional($._statements)),
+        'end',
+      )
+    },
+
+    fun_def: $ => {
+      const params = seq(
+        '(',
+        optional(field('params',
+          alias(choice($.fun_param_list, $.fun_type_param_list), $.param_list),
+        )),
+        ')',
+      )
+
+      const real_name = seq('=',
+        field('real_name', choice($.identifier, $.constant, $.string)),
+      )
+      const return_type = field('type', seq(/[ \t]:\s/, $._bare_type))
+
+      return seq(
+        'fun',
+        field('name', choice($.identifier, $.constant)),
+        optional(real_name),
+        optional(params),
+        optional(return_type),
+      )
+    },
+
+    fun_param_list: $ => {
+      return seq(
+        $.fun_param,
+        repeat(seq(',', $.fun_param)),
+        optional(seq(
+          ',', optional('...'),
+        )),
+      )
+    },
+
+    fun_type_param_list: $ => {
+      return seq(
+        $._type,
+        repeat(seq(',', $._type)),
+        optional(seq(
+          ',', optional('...'),
+        )),
+      )
+    },
+
+    fun_param: $ => {
+      const type = field('type', seq(/[ \t]:\s/, $._bare_type))
+
+      return seq(
+        choice($.identifier, $.constant),
+        type,
+      )
+    },
+
+    type_def: $ => seq(
+      'type',
+      $.constant,
+      '=',
+      $._bare_type,
+    ),
+
+    c_struct_or_union_def: $ => {
```

Comment:
I'm leaning towards splitting this rule. Structs and unions are different things with different semantics, they should have different nodes as well.

Reply:
Can you split all the `c_struct_or_union_*` rules into `struct` and `union` versions?

Reply:
I'm fine splitting `c_struct_or_union_def` into `c_struct_def` and `union_def`. I'm just not sure about the other `_c_struct_or_union_*` rules, as (from the perspective of the parser) they're the same except for the keyword to start them.

Reply:
Makes more sense to have `union_fields` inside a `union_def` than `c_struct_or_union_fields`

Reply:
The Crystal compiler doesn't make this distinction for this reason, and it does something similar with classes/structs, so it should be fine in this case.

Reply:
I could have been more explicit about my thinking on this: the `c_struct_or_union` name seemed unwieldy, and I was also considering that an end user of the syntax tree could potentially want to distinguish union fields from struct fields. (For highlighting, or other editor operations)

After checking the highlight query documentation, I see that maybe duplicating the field types wasn't necessary. I think `(union_def (c_struct_or_union_fields))` and `(struct_def (c_struct_or_union_fields))` patterns would distinguish those fields correctly.

Not a big deal, I just wanted to mention because there could be other parser changes we need to make to support highlighting, etc. And I think we won't know the requirements for those changes until we're writing the query patterns.